### PR TITLE
Add managed istio feature gates for Seed clusters.

### PR DESCRIPTION
**Pull request by mvladev**
**What this PR does / why we need it**:

A customized installation of Istio can now be enabled and installed by gardenlet. The feature gate is `ManagedIstio` and can be configured in the component config.

Disabling this feature once installed does not removes any installed resources as it might accidentally remove already existing installation of Istio.

The installation deploys istiod in the `istio-system` namespace and istio ingress gateway in `istio-ingress` namespace. mTLS is enforced and `Services`, `VirtualServices` and `DestinationRules` are not exported and advertised by default and therefore it must be explicitly enabled either via the `networking.istio.io/exportTo: "*"` annotation or `.exportTo: ["*"]`.

Telemetry is disabled.

**Which issue(s) this PR fixes**:

Part of #1702 

**Special notes for your reviewer**:

Some of the tests are commented out as the `istio.io/api` and `istio.io/client-go` require K8S version `1.18`.

**Release note**:

```noteworthy operator
Added new feature gate `ManagedIstio` to gardenlet. When enabled it deploys a customized installation of [istio](https://istio.io) on Seed clusters. Disabling it once enabled does not remove any installed resources.

The installation deploys istiod in the `istio-system` namespace and istio ingress gateway in `istio-ingress` namespace. mTLS is enforced and `Services`, `VirtualServices` and `DestinationRules` are not exported and advertised by default and therefore it must be explicitly enabled either via the `networking.istio.io/exportTo: "*"` annotation or `.exportTo: ["*"]`.
```

/cc @zanetworker

### CODE REVIEW

File: charts/istio/istio-ingress/templates/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+          periodSeconds: 2
+          successThreshold: 1
+          timeoutSeconds: 1
+        resources:
+          limits:
+            cpu: 2000m
+            memory: 1024Mi
+          requests:
+            cpu: 100m
+            memory: 128Mi
```
Comment (rfranzke) [inline charts/istio/istio-ingress/templates/deployment.yaml]: Should we configure a VPA for this deployment?

File: charts/istio/istio-istiod/templates/configmap-envoy.yaml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      - name: "in.15011"
+        address:
+          socket_address:
+            address: 0.0.0.0
+            port_value: 15011
+        filter_chains:
+        - filters:
+          - name: envoy.http_connection_manager
+            #typed_config
+            #"@type": "type.googleapis.com/",
```
Comment (rfranzke) [inline charts/istio/istio-istiod/templates/configmap-envoy.yaml]: What does this mean?

File: charts/istio/istio-istiod/templates/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        - name: VALIDATION_WEBHOOK_CONFIG_NAME
+          value: istiod
+        - name: PILOT_EXTERNAL_GALLEY
+          value: "false"
+        - name: CLUSTER_ID
+          value: "Kubernetes"
+        resources:
+          requests:
+            cpu: 500m
+            memory: 2048Mi
```
Comment (rfranzke) [inline charts/istio/istio-istiod/templates/deployment.yaml]: What about resource limits? Would a VPA make sense here?

File: docs/deployment/feature_gates.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
+# Feature gates in Gardener
```
Comment (rfranzke) [inline docs/deployment/feature_gates.md]:
```suggestion
# Feature Gates in Gardener
```

File: pkg/features/features.go [hunk 5]
```diff
@@ -33,11 +33,18 @@ const (
...
-	// alpha: v0.1.0
+	// alpha: v0.32.0
 	HVPAForShootedSeed featuregate.Feature = "HVPAForShootedSeed"
+
+	// ManagedIstio installs minimal Istio components in istio-system.
+	// Disable this feature if Istio is already installed in the cluster.
+	// Istio is not automatically removed if this feature is set to false.
+	// owner @mvladev
+	// alpha: v1.4.0
+	ManagedIstio featuregate.Feature = "ManagedIstio"
```
Comment (rfranzke) [inline pkg/features/features.go:50]: Can you also add a document that describes what you get when this feature gate is enabled? What is such a "minimal istio installation", why is it useful and when would I want to enable it?

File: pkg/operation/seed/istio/ingress_gateway.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+		namespace:    namespace,
+		ChartApplier: applier,
+		chartPath:    filepath.Join(chartsRootPath, istioReleaseName, "istio-ingress"),
+		client:       client,
+	}
+}
+
+func (i *ingress) Deploy(ctx context.Context) error {
+	if err := i.client.Create(
+		context.TODO(),
```
Comment (rfranzke) [inline pkg/operation/seed/istio/ingress_gateway.go]:
```suggestion
		ctx,
```

File: pkg/operation/seed/istio/istiod.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+		namespace:    namespace,
+		ChartApplier: applier,
+		chartPath:    filepath.Join(chartsRootPath, istioReleaseName, "istio-istiod"),
+		client:       client,
+	}
+}
+
+func (i *istiod) Deploy(ctx context.Context) error {
+	if err := i.client.Create(
+		context.TODO(),
```
Comment (rfranzke) [inline pkg/operation/seed/istio/istiod.go]:
```suggestion
		ctx,
```

File: pkg/operation/seed/seed.go [hunk 8]
```diff
@@ -601,6 +606,45 @@ func BootstrapCluster(k8sGardenClient kubernetes.Interface, seed *Seed, config *
...
+			common.IstioIngressGatewayNamespace,
+			chartApplier,
+			"charts",
+			k8sSeedClient.Client(),
+		)
+
+		if err := component.OpWaiter(istioCRDs, istiod, igw).Deploy(context.TODO()); err != nil {
+			return err
+		}
+	}
```
Comment (rfranzke) [inline pkg/operation/seed/seed.go:662]: cc @tim-ebert - this should probably be adapted to a dedicated `ManagedResource` as well after your PR is merged

**Review by rfranzke — CHANGES_REQUESTED** [review]

**Comment by rfranzke** [comment]
/kind/enhancement

File: charts/istio/istio-ingress/templates/deployment.yaml (hunk 1 above)
Comment (mvladev) [reply charts/istio/istio-ingress/templates/deployment.yaml]: I'll rather add `HPA` as IGW needs to scale horizontally.

File: charts/istio/istio-istiod/templates/configmap-envoy.yaml (hunk 2 above)
Comment (mvladev) [reply charts/istio/istio-istiod/templates/configmap-envoy.yaml]: Added accidentally.

**Comment by mvladev** [comment]
Added documentation and HPA

File: charts/istio/istio-ingress/templates/deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        - name: istio-token
+          mountPath: /var/run/secrets/tokens
+          readOnly: true
+        - name: ingressgatewaysdsudspath
+          mountPath: /var/run/ingress_gateway
+        - name: podinfo
+          mountPath: /etc/istio/pod
+      volumes:
+      - name: istiod-ca-cert
+        configMap:
```
Comment (rfranzke) [inline charts/istio/istio-ingress/templates/deployment.yaml:141]: Do we need a checksum annotation for this config map?

File: charts/istio/istio-istiod/templates/deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+          sources:
+            - serviceAccountToken:
+                audience: istio-ca
+                expirationSeconds: 43200
+                path: istio-token
+      - name: config-volume
+        configMap:
+          name: istio
+      - name: pilot-envoy-config
+        configMap:
```
Comment (rfranzke) [inline charts/istio/istio-istiod/templates/deployment.yaml]: The istio config already has a checksum annotation, do we need a checksum annotation for this config map?

File: docs/concepts/istio.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
+# Istio
+
+[Istio](https://istio.io) offers a service mesh implementation with focus on several important features - traffic, observability, security and policy.
+
+## Gardener ManagedIstio feature gate
```
Comment (rfranzke) [inline docs/concepts/istio.md]:
```suggestion
## Gardener `ManagedIstio` feature gate
```

File: docs/concepts/istio.md [hunk 12]
```diff
@@ -0,0 +1,26 @@
+# Istio
```
Comment (rfranzke) [inline docs/concepts/istio.md]: Can you move the docs to the `docs/usage` folder? Can you also add links to them in the index (`docs/README.md`).

File: docs/concepts/istio.md [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+
+[Istio](https://istio.io) offers a service mesh implementation with focus on several important features - traffic, observability, security and policy.
+
+## Gardener ManagedIstio feature gate
+
+When enabled in gardenlet the `ManagedIstio` feature gate can be used to deploy a Gardener-tailored Istio installation in Seed clusters. It's main usage is to enable features such as [Shoot API server SNI](proposals/08-shoot-apiserver-via-sni.md). This feature should not be enabled on a Seed cluster which already deploy Istio.
```
Comment (rfranzke) [inline docs/concepts/istio.md]:
```suggestion
When enabled in gardenlet the `ManagedIstio` feature gate can be used to deploy a Gardener-tailored Istio installation in Seed clusters. It's main usage is to enable features such as [Shoot API server SNI](proposals/08-shoot-apiserver-via-sni.md). This feature should not be enabled on a Seed cluster where Istio is already deployed.
```

File: docs/concepts/istio.md [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+## Prerequisites
+
+Third-party JWT is used, therefore each Seed cluster where this feature is enabled must have [Service Account Token Volume Projection](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection) enabled.
+
+## Differences with Istio's default profile
+
+The [default profile](https://istio.io/docs/setup/additional-setup/config-profiles/) which is recommended for production deployment, is not suitable for the Gardener use case as it offers more functionality that's desired. The current installation goes through heavy refactoring due to the `IstioOperator` and the mixture of Helm values + Kubernetes API specification makes configuring and fine-tuning it very hard. A more simplistic deployment is used by Gardener. The differences are the following:
```
Comment (rfranzke) [inline docs/concepts/istio.md]:
```suggestion
The [default profile](https://istio.io/docs/setup/additional-setup/config-profiles/) which is recommended for production deployment, is not suitable for the Gardener use case as it offers more functionality than desired. The current installation goes through heavy refactorings due to the `IstioOperator` and the mixture of Helm values + Kubernetes API specification makes configuring and fine-tuning it very hard. A more simplistic deployment is used by Gardener. The differences are the following:
```

File: docs/concepts/istio.md [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+- Telemetry is not deployed.
+- `istiod` is deployed.
+- `istio-ingress-gateway` is deployed in a separate `istio-ingress` namespace.
+- `istio-egress-gateway` is not deployed.
+- None of the Istio addons are deployed.
+- Mixer (deprecated) is not deployed
+- Mixer CDRs are not deployed.
+- Kubernetes `Service`, Istio's `VirtualService` and `ServiceEntry` are **NOT** advertised in the service mesh. This means that if a `Service` needs to be accessed directly from the Istio Ingress Gateway, it should have `networking.istio.io/exportTo: "*"` annotation. `VirtualService` and `ServiceEntry` must have `.spec.exportTo: ["*"]` set on them respectively.
+- Istio injector is not enabled.
+- mTLS is enabled by default.
```
Comment (rfranzke) [inline docs/concepts/istio.md]: Nice summary, thank you!

File: pkg/features/features.go [hunk 16]
```diff
@@ -33,11 +33,19 @@ const (
...
-	// alpha: v0.1.0
+	// alpha: v0.32.0
 	HVPAForShootedSeed featuregate.Feature = "HVPAForShootedSeed"
+
+	// ManagedIstio installs minimal Istio components in istio-system.
+	// Disable this feature if Istio is already installed in the cluster.
+	// Istio is not automatically removed if this feature is set to false.
+	// See https://github.com/gardener/gardener/blob/master/docs/concepts/istio.md
+	// owner @mvladev
+	// alpha: v1.4.0
```
Comment (rfranzke) [inline pkg/features/features.go]:
```suggestion
	// alpha: v1.5.0
```

File: charts/istio/istio-ingress/templates/deployment.yaml (hunk 9 above)
Comment (mvladev) [reply charts/istio/istio-ingress/templates/deployment.yaml:141]: No, it's reloaded automatically.

File: charts/istio/istio-istiod/templates/deployment.yaml (hunk 10 above)
Comment (mvladev) [reply charts/istio/istio-istiod/templates/deployment.yaml]: It's a leftover

**Comment by mvladev** [comment]
Updated the PR - have a look again.

**Comment by rfranzke** [comment]
When trying it out I get following error:

```
time="2020-05-13T08:03:06+02:00" level=error msg="Seed bootstrapping failed: failed to apply manifests: 1 error occurred: could not read object: error 'error converting YAML to JSON: yaml: line 20: mapping values are not allowed in this context' decoding manifest: \n---\n# Source: istio-istiod/templates/configmap-envoy.yaml\napiVersion: v1\nkind: ConfigMap\nmetadata:\n  name: pilot-envoy-config\n  namespace: istio-system\n  labels:\n    app: istiod\n    istio: pilot\n    \ndata:\n  envoy.yaml.tmpl: |-\n    admin:\n      access_log_path: /dev/null\n      address:\n        socket_address:\n          address: 127.0.0.1\n          port_value: 15000\n\n    static_resources:\n      clusters:\n      - name: in.15010\n        http2_protocol_options: {}\n        connect_timeout: 1.000s\n\n        hosts:\n        - socket_address:\n            address: 127.0.0.1\n            port_value: 15010\n\n        circuit_breakers:\n          thresholds:\n          - max_connections: 100000\n            max_pending_requests: 100000\n            max_requests: 100000\n            max_retries: 3\n\n      listeners:\n      - name: \"in.15011\"\n        address:\n          socket_address:\n            address: 0.0.0.0\n            port_value: 15011\n        filter_chains:\n        - filters:\n          - name: envoy.http_connection_manager\n            config:\n              codec_type: HTTP2\n              stat_prefix: \"15011\"\n              stream_idle_timeout: 0s\n              http2_protocol_options:\n                max_concurrent_streams: 1073741824\n\n              access_log:\n              - name: envoy.file_access_log\n                config:\n                  path: /dev/stdout\n\n              http_filters:\n              - name: envoy.router\n\n              route_config:\n                name: \"15011\"\n\n                virtual_hosts:\n                - name: istio-pilot\n\n                  domains:\n                  - '*'\n\n                  routes:\n                  - match:\n                      prefix: /\n                    route:\n                      cluster: in.15010\n                      timeout: 0.000s\n                    decorator:\n                      operation: xDS\n\n          tls_context:\n            require_client_certificate: true\n            common_tls_context:\n              validation_context:\n                trusted_ca:\n                  filename: /etc/certs/root-cert.pem\n\n              alpn_protocols:\n              - h2\n\n              tls_certificates:\n              - certificate_chain:\n                  filename: /etc/certs/cert-chain.pem\n                private_key:\n                  filename: /etc/certs/key.pem\n\n---\n# Source: istio-istiod/templates/configmap.yaml\napiVersion: v1\nkind: ConfigMap\nmetadata:\n  name: istio\n  namespace: istio-system\n  labels:\n    app: istiod\n    istio: pilot\n    \ndata:\n\n  # Configuration file for the mesh networks to be used by the Split Horizon EDS.\n  meshNetworks: |-\n    networks: {}\n\n  mesh: |-\n    # Set enableTracing to false to disable request tracing.\n    enableTracing: false\n\n    # Set accessLogFile to empty string to disable access log.\n    accessLogFile: \"\"\n\n    accessLogFormat: \"\"\n\n    accessLogEncoding: 'TEXT'\n\n    enableEnvoyAccessLogService: false\n    # reportBatchMaxEntries is the number of requests that are batched before telemetry data is sent to the mixer server\n    reportBatchMaxEntries: 100\n    # reportBatchMaxTime is the max waiting time before the telemetry data of a request is sent to the mixer server\n    reportBatchMaxTime: 1s\n    disableMixerHttpReports: false\n\n    # Set the following variable to true to disable policy checks by the Mixer.\n    # Note that metrics will still be reported to the Mixer.\n    disablePolicyChecks: true\n\n    # Automatic protocol detection uses a set of heuristics to\n    # determine whether the connection is using TLS or not (on the\n    # server side), as well as the application protocol being used\n    # (e.g., http vs tcp). These heuristics rely on the client sending\n    # the first bits of data. For server first protocols like MySQL,\n    # MongoDB, etc., Envoy will timeout on the protocol detection after\n    # the specified period, defaulting to non mTLS plain TCP\n    # traffic. Set this field to tweak the period that Envoy will wait\n    # for the client to send the first bits of data. (MUST BE >=1ms)\n    protocolDetectionTimeout: 100ms\n\n    # This is the k8s ingress service name, not used.\n    ingressService: \"istio-ingressgateway\"\n    ingressControllerMode: \"OFF\"\n    ingressClass: \"istio\"\n\n    # The trust domain corresponds to the trust root of a system.\n    # Refer to https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE-ID.md#21-trust-domain\n    trustDomain: \"cluster.local\"\n\n    # Used by pilot-agent\n    sdsUdsPath: \"unix:/etc/istio/proxy/SDS\"\n\n    # If true, automatically configure client side mTLS settings to match the corresponding service's\n    # server side mTLS authentication policy, when destination rule for that service does not specify\n    # TLS settings.\n    enableAutoMtls: false\n\n    outboundTrafficPolicy:\n      # allow traffic only to services in the mesh\n      mode: REGISTRY_ONLY\n    localityLbSetting:\n      enabled: true\n\n    # Configures DNS certificates provisioned through Chiron linked into Pilot.\n    # The DNS certificate provisioning is enabled by default now so it get tested.\n    certificates: []\n\n    # Disable the advertisment of services and endpoints which are no explictly marked in\n    # `exportTo`. Improves security and isolation.\n    # Refer to https://istio.io/docs/reference/config/istio.mesh.v1alpha1/#MeshConfig\n    defaultServiceExportTo: [\"-\"]\n    defaultVirtualServiceExportTo: [\"-\"]\n    defaultDestinationRuleExportTo: [\"-\"]\n\n    defaultConfig:\n      #\n      # TCP connection timeout between Envoy & the application, and between Envoys.\n      connectTimeout: 10s\n      #\n      ### ADVANCED SETTINGS #############\n      # Where should envoy's configuration be stored in the istio-proxy container\n      configPath: \"/etc/istio/proxy\"\n      # The pseudo service name used for Envoy.\n      serviceCluster: istio-proxy\n      # These settings that determine how long an old Envoy\n      # process should be kept alive after an occasional reload.\n      drainDuration: 45s\n      parentShutdownDuration: 1m0s\n      #\n      # Port where Envoy listens (on local host) for admin commands\n      # You can exec into the istio-proxy container in a pod and\n      # curl the admin port (curl http://localhost:15000/) to obtain\n      # diagnostic information from Envoy. See\n      # https://lyft.github.io/envoy/docs/operations/admin.html\n      # for more details\n      proxyAdminPort: 15000\n      #\n      # Set concurrency to a specific number to control the number of Proxy worker threads.\n      # If set to 0 (default), then start worker thread for each CPU thread/core.\n      concurrency: 2\n      #\n      # If port is 15012, will use SDS.\n      # controlPlaneAuthPolicy is for mounted secrets, will wait for the files.\n      controlPlaneAuthPolicy: NONE\n      discoveryAddress: istiod.istio-system.svc:15012\n\n---\n# Source: istio-istiod/templates/serviceaccount.yaml\napiVersion: v1\nkind: ServiceAccount\nmetadata:\n  name: istiod\n  namespace: istio-system\n  labels:\n    app: istiod\n    istio: pilot\n    \n\n---\n# Source: istio-istiod/templates/clusterrole.yaml\napiVersion: rbac.authorization.k8s.io/v1\nkind: ClusterRole\nmetadata:\n  name: istiod\n  labels:\n    app: istiod\n    istio: pilot\n    \nrules:\n# sidecar injection controller Do we need it?\n- apiGroups:\n  - admissionregistration.k8s.io\n  resources:\n  - mutatingwebhookconfigurations\n  verbs:\n  - get\n  - list\n  - watch\n  - patch\n# configuration validation webhook controller\n- apiGroups:\n  - admissionregistration.k8s.io\n  resources:\n  - validatingwebhookconfigurations\n  verbs:\n  - get\n  - list\n  - watch\n  - update\n# permissions to verify the webhook is ready and rejecting\n# invalid config. We use --server-dry-run so no config is persisted.\n- apiGroups:\n  - networking.istio.io\n  verbs:\n  - create\n  resources:\n  - gateways\n# istio configuration\n- apiGroups:\n  - config.istio.io\n  - rbac.istio.io\n  - security.istio.io\n  - networking.istio.io\n  - authentication.istio.io\n  verbs:\n  - get\n  - watch\n  - list\n  resources:\n  - \"*\"\n# auto-detect installed CRD definitions\n- apiGroups:\n  - apiextensions.k8s.io\n  resources:\n  - customresourcedefinitions\n  verbs:\n  - get\n  - list\n  - watch\n# discovery and routing\n- apiGroups:\n  - extensions\n  - apps\n  resources:\n  - deployments\n  verbs:\n  - get\n  - list\n  - watch\n# discovery and routing\n- apiGroups:\n  - ''\n  resources:\n  - pods\n  - nodes\n  - services\n  - namespaces\n  - endpoints\n  verbs:\n  - get\n  - list\n  - watch\n# discovery and routing\n- apiGroups:\n  - discovery.k8s.io\n  resources:\n  - endpointslices\n  verbs:\n  - get\n  - list\n  - watch\n# ingress controller TODO: is this needed???\n- apiGroups:\n  - extensions\n  resources:\n  - ingresses\n  verbs:\n  - get\n  - list\n  - watch\n- apiGroups:\n  - extensions\n  resources:\n  - ingresses/status\n  verbs:\n  - \"*\"\n# required for CA's namespace controller\n- apiGroups:\n  - ''\n  resources:\n  - configmaps\n  verbs:\n  - create\n  - get\n  - list\n  - watch\n  - update\n# Istiod and bootstrap.\n- apiGroups:\n  - certificates.k8s.io\n  resources:\n  - certificatesigningrequests\n  - certificatesigningrequests/approval\n  - certificatesigningrequests/status\n  verbs:\n  - update\n  - create\n  - get\n  - delete\n  - watch\n# Used by Istiod to verify the JWT tokens\n- apiGroups:\n  - authentication.k8s.io\n  resources:\n  - tokenreviews\n  verbs:\n  - create\n# TODO: remove, no longer needed at cluster\n- apiGroups:\n  - ''\n  resources:\n  - secrets\n  verbs:\n  - create\n  - get\n  - watch\n  - list\n  - update\n  - delete\n# TODO: remove, no longer needed at cluster\n- apiGroups:\n  - ''\n  resources:\n  - serviceaccounts\n  verbs:\n  - get\n  - watch\n  - list\n\n---\n# Source: istio-istiod/templates/clusterrolebinding.yaml\napiVersion: rbac.authorization.k8s.io/v1\nkind: ClusterRoleBinding\nmetadata:\n  name: istiod\n  labels:\n    app: istiod\n    istio: pilot\n    \nroleRef:\n  apiGroup: rbac.authorization.k8s.io\n  kind: ClusterRole\n  name: istiod\nsubjects:\n- kind: ServiceAccount\n  name: istiod\n  namespace: istio-system\n\n---\n# Source: istio-istiod/templates/role.yaml\napiVersion: rbac.authorization.k8s.io/v1\nkind: Role\nmetadata:\n  name: istiod\n  namespace: istio-system\n  labels:\n    app: istiod\n    istio: pilot\n    \nrules:\n- apiGroups:\n  - ''\n  resources:\n  - secrets\n  verbs:\n  - create\n  - get\n  - watch\n  - list\n  - update\n  - delete\n- apiGroups:\n  - ''\n  resources:\n  - serviceaccounts\n  verbs:\n  - get\n  - watch\n  - list\n\n---\n# Source: istio-istiod/templates/rolebinding.yaml\napiVersion: rbac.authorization.k8s.io/v1\nkind: RoleBinding\nmetadata:\n  name: istiod\n  labels:\n    app: istiod\n    istio: pilot\n    \nroleRef:\n  apiGroup: rbac.authorization.k8s.io\n  kind: Role\n  name: istiod\nsubjects:\n- kind: ServiceAccount\n  name: istiod\n\n---\n# Source: istio-istiod/templates/service.yaml\napiVersion: v1\nkind: Service\nmetadata:\n  name: istiod\n  namespace: istio-system\n  labels:\n    app: istiod\n    istio: pilot\n    \nspec:\n  ports:\n  - name: https-sds # mTLS with k8s-signed cert\n    port: 15012\n  - name: https-webhook # validation and injection\n    port: 443\n    targetPort: 15017\n  selector:\n    app: istiod\n    istio: pilot\n    \n\n---\n# Source: istio-istiod/templates/deployment.yaml\napiVersion: apps/v1\nkind: Deployment\nmetadata:\n  name: istiod\n  namespace: istio-system\n  labels:\n    app: istiod\n    istio: pilot\n    \nspec:\n  strategy:\n    rollingUpdate:\n      maxSurge: 100%\n      maxUnavailable: 25%\n  selector:\n    matchLabels:\n      app: istiod\n      istio: pilot\n      \n  template:\n    metadata:\n      labels:\n        app: istiod\n        istio: pilot\n        \n      annotations:\n        sidecar.istio.io/inject: \"false\"\n        checksum/istio-config: 353bdf30a5e48727d1659ea4d655376fb881ea55d8ba375fa923f8428807c71e\n    spec:\n      serviceAccountName: istiod\n      securityContext:\n        fsGroup: 1337\n      containers:\n      - name: discovery\n        image: \"docker.io/istio/pilot:1.5.2\"\n        imagePullPolicy: IfNotPresent\n        args:\n        - \"discovery\"\n        - --monitoringAddr=\n        - --secureGrpcAddr=\n        - --grpcAddr=\n        - --httpAddr=localhost:8080\n        - --httpsAddr=:15017\n        - --log_output_level=all:warn,ads:error\n        - --domain=cluster.local\n        - --trust-domain=cluster.local\n        - --keepaliveMaxServerConnectionAge=30m\n        ports:\n        - containerPort: 15012\n        - containerPort: 15017\n        readinessProbe:\n          exec:\n            command:\n            - curl\n            - -s\n            - http://localhost:8080/ready\n          initialDelaySeconds: 5\n          periodSeconds: 5\n          timeoutSeconds: 5\n        env:\n        - name: JWT_POLICY\n          value: third-party-jwt\n        - name: PILOT_CERT_PROVIDER\n          value: istiod\n        - name: POD_NAME\n          valueFrom:\n            fieldRef:\n              apiVersion: v1\n              fieldPath: metadata.name\n        - name: POD_NAMESPACE\n          valueFrom:\n            fieldRef:\n              apiVersion: v1\n              fieldPath: metadata.namespace\n        - name: SERVICE_ACCOUNT\n          valueFrom:\n            fieldRef:\n              apiVersion: v1\n              fieldPath: spec.serviceAccountName\n        - name: PILOT_TRACE_SAMPLING\n          value: \"0.1\"\n        - name: PILOT_ENABLE_PROTOCOL_SNIFFING_FOR_OUTBOUND\n          value: \"false\"\n        - name: PILOT_ENABLE_PROTOCOL_SNIFFING_FOR_INBOUND\n          value: \"false\"\n        - name: INJECTION_WEBHOOK_CONFIG_NAME\n          value: istio-sidecar-injector\n        - name: ISTIOD_ADDR\n          value: istiod.istio-system.svc:15012\n        - name: VALIDATION_WEBHOOK_CONFIG_NAME\n          value: istiod\n        - name: PILOT_EXTERNAL_GALLEY\n          value: \"false\"\n        - name: CLUSTER_ID\n          value: \"Kubernetes\"\n        resources:\n          requests:\n            cpu: 250m\n            memory: 128Mi\n          limits:\n            cpu: 1000m\n            memory: 2048Mi\n        securityContext:\n          runAsUser: 1337\n          runAsGroup: 1337\n          runAsNonRoot: true\n          capabilities:\n            drop:\n            - ALL\n        volumeMounts:\n        - name: config-volume\n          mountPath: /etc/istio/config\n        - name: istio-token\n          mountPath: /var/run/secrets/tokens\n          readOnly: true\n        - name: local-certs\n          mountPath: /var/run/secrets/istio-dns\n      volumes:\n      # Technically not needed on this pod - but it helps debugging/testing SDS\n      # Should be removed after everything works.\n      - emptyDir:\n          medium: Memory\n        name: local-certs\n      - name: istio-token\n        projected:\n          sources:\n            - serviceAccountToken:\n                audience: istio-ca\n                expirationSeconds: 43200\n                path: istio-token\n      - name: config-volume\n        configMap:\n          name: istio\n      affinity:\n        podAntiAffinity:\n          preferredDuringSchedulingIgnoredDuringExecution:\n          - weight: 100\n            podAffinityTerm:\n              labelSelector:\n                matchExpressions:\n                - key: app\n                  operator: In\n                  values:\n                  - istiod\n                - key: istio\n                  operator: In\n                  values:\n                  - pilot\n              topologyKey: \"kubernetes.io/hostname\"\n\n---\n# Source: istio-istiod/templates/destinationrule.yaml\napiVersion: networking.istio.io/v1beta1\nkind: DestinationRule\nmetadata:\n  name: default\n  namespace: istio-system\nspec:\n  host: \"*\"\n  exportTo:\n  - \"*\"\n  trafficPolicy:\n    tls:\n      mode: ISTIO_MUTUAL\n\n---\n# Source: istio-istiod/templates/autoscale.yaml\napiVersion: autoscaling/v2beta1\nkind: HorizontalPodAutoscaler\nmetadata:\n  name: istiod\n  namespace: istio-system\n  labels:\n    app: istiod\n    istio: pilot\n    \nspec:\n  maxReplicas: 2\n  minReplicas: 5\n  scaleTargetRef:\n    apiVersion: apps/v1\n    kind: Deployment\n    name: istiod\n  metrics:\n  - type: Resource\n      resource:\n      name: cpu\n      targetAverageUtilization: 80\n\n---\n# Source: istio-istiod/templates/peerauthentication.yaml\napiVersion: security.istio.io/v1beta1\nkind: PeerAuthentication\nmetadata:\n  name: default\n  namespace: istio-system\nspec:\n  mtls:\n    mode: STRICT\n\n---\n# Source: istio-istiod/templates/sidecar.yaml\napiVersion: networking.istio.io/v1alpha3\nkind: Sidecar\nmetadata:\n  name: default\n  namespace: istio-system\n  labels:\n    app: istiod\n    istio: pilot\n    \nspec:\n  egress:\n  - hosts:\n    - \"*/*\"\n  outboundTrafficPolicy:\n    mode: REGISTRY_ONLY\n\n---\n# Source: istio-istiod/templates/namespace.yaml\n\n\n---\n# Source: istio-istiod/templates/validatingwebhookconfiguration.yaml\napiVersion: admissionregistration.k8s.io/v1beta1\nkind: ValidatingWebhookConfiguration\nmetadata:\n  name: istiod\n  labels:\n    app: istiod\n    istio: pilot\n    \nwebhooks:\n  - name: validation.istio.io\n    clientConfig:\n      service:\n        name: istiod\n        namespace: istio-system\n        path: \"/validate\"\n      caBundle: \"\" # patched at runtime when the webhook is ready.\n    rules:\n      - operations:\n        - CREATE\n        - UPDATE\n        apiGroups:\n        - config.istio.io\n        - rbac.istio.io\n        - security.istio.io\n        - authentication.istio.io\n        - networking.istio.io\n        apiVersions:\n        - \"*\"\n        resources:\n        - \"*\"\n    # Fail open until the validation webhook is ready. The webhook controller\n    # will update this to `Fail` and patch in the `caBundle` when the webhook\n    # endpoint is ready.\n    failurePolicy: Ignore\n    sideEffects: None\n" seed=aws
```

**Comment by rfranzke** [comment]
/status/author-action

**Comment by rfranzke** [comment]
Can you also squash the commits?

**Comment by mvladev** [comment]
All istio related commits are now one. The `move FakeDiscovery in own package` commit is still a separate one.

**Review by rfranzke — CHANGES_REQUESTED** [review]
```
time="2020-05-14T10:22:52+02:00" level=error msg="Seed bootstrapping failed: failed to apply manifests: 2 errors occurred: [could not apply object of kind \"DestinationRule\" \"istio-system/default\": Internal error occurred: no kind \"DestinationRule\" is registered for version \"networking.istio.io/v1beta1\" in scheme \"k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:51\", could not apply object of kind \"HorizontalPodAutoscaler\" \"istio-system/istiod\": HorizontalPodAutoscaler.autoscaling \"istiod\" is invalid: spec.maxReplicas: Invalid value: 2: must be greater than or equal to `minReplicas`]" seed=aws
```

/needs/changes

**Comment by rfranzke** [comment]
/area/networking
/size/xl

**Comment by mvladev** [comment]
Rebased to latest changes

**Review by rfranzke — CHANGES_REQUESTED** [review]
@mvladev I still get errors when I try it locally. Does it work for you?

```
time="2020-05-22T07:18:53+02:00" level=error msg="Seed bootstrapping failed: failed to apply manifests: 1 error occurred: could not apply object of kind \"DestinationRule\" \"istio-system/default\": Internal error occurred: no kind \"DestinationRule\" is registered for version \"networking.istio.io/v1beta1\" in scheme \"k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:51\"" seed=aws
```

Probably you have to add the Istio API types to the [`SeedScheme`](https://github.com/gardener/gardener/blob/master/pkg/client/kubernetes/types.go#L73-L80)?

**Review by rfranzke — APPROVED** [review]
/lgtm
/remove status author-action

Turns out the problem was probably caused by an earlier version of this PR when I tried it out the first time. Removing the CRD in question fixed it.

**Pull request merged**